### DISCUSSION TITLE
Sourcegraph precise code navigation with scala-cli

### DISCUSSION THREAD

**Discussion by ckipp01** [VirtusLab/scala-cli#549] [Show and tell]
I'm a big [Sourcegraph](https://sourcegraph.com/) fan. I use it both when I'm searching for code, and also on GitHub as a browser extension to be able to navigate around my code more intelligently. I was able to set up indexing with my scala-cli project enabling precise code navigation which I think is pretty neat.

![2022-01-17 10 17 22](https://user-images.githubusercontent.com/13974112/149741881-09d16291-7737-468c-84ab-37587ba39718.gif)

You can see an example of how I set this up [here](https://github.com/ckipp01/exists/blob/main/.github/workflows/CI.yml).

One missing piece that I'd love to have though is cross-repository navigation. In order to achieve this you need to be able to grab what is actually passed to the compiler (basically most interested in the classpath), and put that in a file to also be uploaded. This is outlined [here](https://sourcegraph.github.io/lsif-java/docs/manual-configuration.html#step-5-optional-enable-cross-repository-navigation). I can sort of see this information when I use a couple `-v -v -v` during the compile, but is there any way for me to actually grab this information or export this information? I basically just want this info:

```
Fetching List(org.scala-lang::scala3-compiler:3.1.0), adding List(ivy:file:///Users/ckipp/Library/Caches/ScalaCli/local-repo/v0.0.9//[defaultPattern])
Found 13 artifacts:
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/3.1.0/scala3-compiler_3-3.1.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-interfaces/3.1.0/scala3-interfaces-3.1.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.1.0/scala3-library_3-3.1.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/tasty-core_3/3.1.0/tasty-core_3-3.1.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-asm/9.1.0-scala-1/scala-asm-9.1.0-scala-1.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/compiler-interface/1.3.5/compiler-interface-1.3.5.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-reader/3.19.0/jline-reader-3.19.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-terminal/3.19.0/jline-terminal-3.19.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-terminal-jna/3.19.0/jline-terminal-jna-3.19.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.6/scala-library-2.13.6.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/google/protobuf/protobuf-java/3.7.0/protobuf-java-3.7.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/util-interface/1.3.0/util-interface-1.3.0.jar
  /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar

Fetching List(org.scala-lang::scala3-library::3.1.0, org.jsoup:jsoup:1.14.3, org.scalameta::munit::1.0.0-M1, org.virtuslab.scala-cli::runner:0.0.9)
```
Outputted or available for me to grab it, stick it in the file, and then let lsif-java do it's thing, get the info from their pom files, and then upload them. This would allow for me to jump right to jsoup or the scala standard lib from my project.

Thanks!

**Comment by olafurpg** [comment]
> One missing piece that I'd love to have though is cross-repository navigation.

To elaborate, a JVM dependency is encoded as a "repository" in Sourcegraph terminology and "cross-repository navigation" is the ability to navigate to external dependencies from Maven Central or the JDK. Whenever a user uploads an LSIF file that references a JVM dependency like `org.scala-lang:scala-library:2.12.12` then Sourcegraph fetches the sources of that artifact from Maven Central and creates a repository for it https://sourcegraph.com/maven/org.scala-lang/scala-library
